### PR TITLE
Enable the case search index when turning on new case API flag

### PR DESCRIPTION
**Pull request by esoergel**
## Summary

https://dimagi-dev.atlassian.net/browse/QA-3314

## Feature Flag
Enable the v0.6 Case API

## Product Description
Enable the case search index when turning on new case API flag

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/hqcase/views.py [hunk 1]
```diff
@@ -91,8 +91,12 @@ def delete_cases(self, request, domain):
 def case_api(request, domain, case_id=None):
     if request.method == 'GET' and case_id:
         return _handle_individual_get(request, case_id)
-    if request.method == 'GET' and not case_id and domain_needs_search_index(domain):
-        return _handle_list_view(request)
+    if request.method == 'GET' and not case_id:
+        if domain_needs_search_index(domain):
+            return _handle_list_view(request)
+        return JsonResponse({
+            'error': "You need the case search index to use this feature"
```
Comment (proteusvacuum) [inline corehq/apps/hqcase/views.py:98]: what if you added the `CASE_API_V0_6` flag to this list instead?

https://github.com/dimagi/commcare-hq/blob/ae9f57d878a71def07af3435e27052c20a2b3d04/corehq/pillows/case_search.py#L61-L65

Comment (esoergel) [reply corehq/apps/hqcase/views.py:98]: Initially I left it out thinking that there might be projects that want to use the case update API and not the case list, and we needn't force the case search index on them.  That said, I think adding more projects to the case search is probably in our best interest and might not be a real blocker, so I like that proposal.

Wouldn't I also need some logic to do the initial build of the case search index for the domain?  Like this `save_fn`:
https://github.com/dimagi/commcare-hq/blob/703310fd517e34ffc74ee1b42389f6620f05b960/corehq/toggles.py#L819-L825

It looks like the other feature flags don't do that automatically, although `ECD_MIGRATED_DOMAINS` appears to be part of a mgmt command that turns it on separately.

Comment (proteusvacuum) [reply corehq/apps/hqcase/views.py:98]: Yeah, I think you could piggy back on that save_fn to build the initial index.

File: corehq/toggles.py [hunk 2]
```diff
@@ -808,11 +808,13 @@ def _ensure_valid_randomness(randomness):
...
+    from corehq.apps.es import CaseSearchES
     from corehq.pillows.case_search import domains_needing_search_index
     domains_needing_search_index.clear()
-    if enabled:
-        # action is no longer reversible due to roll out of EXPLORE_CASE_DATA
-        # and upcoming migration of CaseES to CaseSearchES
+
+    has_case_search_cases = CaseSearchES().domain(domain).count() > 0
+    if enabled and not has_case_search_cases:
+        # action is not reversible, we want all projects here eventually
```
Comment (proteusvacuum) [inline corehq/toggles.py:817]: I hadn't seen this change before. Do you think we should remove the `delete_case_search_cases_for_domain` here? (Not for this PR)

https://github.com/dimagi/commcare-hq/blob/8224a32e34d5e8493d9cbc92193334beb68f295c/corehq/apps/case_search/models.py#L177-L192

**Review by proteusvacuum — APPROVED** [review]
Great! Thanks for making those changes.

File: corehq/toggles.py (hunk 2 above)
Comment (esoergel) [reply corehq/toggles.py:817]: Hmm yeah, that's a good point, probably.  And it looks like the `enable_case_search_for_ecd` command might also need an update like what I added here.

https://github.com/dimagi/commcare-hq/blob/703310fd517e34ffc74ee1b42389f6620f05b960/corehq/apps/reports/management/commands/enable_case_search_for_ecd.py#L53-L55

If the `save_fn` doesn't delete the index when the FF is disabled, and we do the same for `disable_case_search`, then there would be migrated domains that have data in the case search index.

_However_, if no feature flags are enabled, then `domains_needing_search_index` won't include that domain, and I assume the pillow won't process new case changes in those domains.  Honestly it sounds like we need a single place to store whether or not the index is enabled, one which isn't tied to user-facing features, and which is never turned off (`ECD_MIGRATED_DOMAINS`?).  Think I might punt that to https://dimagi-dev.atlassian.net/browse/USH-904

Comment (proteusvacuum) [reply corehq/toggles.py:817]: 👍🏽

**Pull request merged**